design-proposals/klip-49-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,129 @@
+# KLIP 47 - Add Source Stream/Table Semantic 
+
+**Author**: Boyang Chen (@boyang) | 
+**Release Target**: TBD | 
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/7474
+
+**tl;dr:** _Add support for source stream/table. This will give users a read-only view of data that 
+are populated by the upstream Kafka topic, making the data ownership more clear._
+           
+## Motivation and background
+
+Users of KsqlDB have no clear ownership definition around tables or streams they created. In a “normal” database, 
+when a user creates a table the DB creates the corresponding files on disk and owns the table. 
+User may write data into those tables via `INSERT INTO` statements and/or modify/delete existing data 
+via `UPDATE` and `DELETE` statements. However, the table will be not changed by any other means. 
+
+For ksqlDB, the setup is quite differently. Specifically, user could create the table based on 
+the topics stored in Kafka as the source of truth. However, since there is no clear ownership, the defined table 
+could be mutated by using `INSERT` command to populate data into it. The insertion will pollute the original data source, 
+which was supposed to be owned by upstream such as a CDC pipeline that populates the topic.
+Secondly, when calling `CREATE TABLE`, the actual data is neither materialized nor ready for interactive queries to access. 
+To address these gaps, we would like to add:
+
+* Syntax to add read-only stream/table as `CREATE SOURCE STREAM/TABLE` to make the data source as read-only. Any 
+insertion command will be rejected, with an example error response as:
+
+```
+Error: insertion into source stream/table is not allowed
+```   
+
+* In addition, the `CREATE SOURCE TABLE` will become a persistent query instead of just a metadata operation like 
+`CREATE TABLE`, `CREATE STREAM`, or `CREATE SOURCE STREAM`. Source table will materialize the input topic data as 
+a RocksDB instance for interactive queries in the KsqlDB servers.
+
+## What is in scope
+* Add `CREATE SOURCE STREAM/TABLE` syntax to KSQL
+* `DESCRIBE ... EXTENDED` should show source streams and tables as read-only  
+* Add materialization of the `source table` to make it pull-query accessible
+* Make source stream/table immutable from KSQL by rejecting insertion requests
+ 
+
+## What is not in scope
+* We will not add role-based write permission for source stream/table in v1, even though it seems to be useful.
+
+## Design
+
+### Semantic Change
+We shall first add optional keyword `SOURCE` into the KSQL codebase, by adding the new syntax to `sqlBase.g4` for:
```

Comment:
nit: no need to change this for this KLIP, but going forward I think we don't need to go into implementation details this deep - it distracts a bit from the design aspect of the KLIP (these kind of changes are better reviewed in PRs)